### PR TITLE
Add Bot read API with production-ready configuration

### PR DESCRIPTION
**Pull request by miller46**
## Summary

Implements read-only API endpoints for the Bot model with comprehensive filtering, pagination, and production-safe configuration.

### Features
- **GET /api/v1/bots** - List bots with pagination, filtering, and sorting
  - Pagination: `page`, `per_page` (max 100)
  - Filters: `rig_id`, `kill_switch`, `log_search` (case-insensitive)
  - Date filters: `create_at_after/before`, `last_update_at_after/before`, `last_run_at_after/before`
  - Sorted by `create_at` DESC with secondary sort by `id` for deterministic ordering

- **GET /api/v1/bots/{id}** - Get single bot by ID
  - Returns 404 with detailed error for missing bots

### Security
- SQL injection prevention: wildcards (`%`, `_`) escaped in log search
- Production database validation: SQLite rejected in production/staging environments

### Architecture
- Database session management with `@lru_cache` for clean dependency injection
- Pydantic response schemas with OpenAPI examples
- Shared test fixtures in `conftest.py`

## Test Coverage
- 39 tests covering all endpoints, filters, pagination, and edge cases
- Deterministic date filter tests with explicit timestamps
- Environment-specific database URL validation tests

## Test plan
- [x] All 39 tests passing
- [x] Ruff linting passes
- [ ] Manual API testing with sample data

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/jm_api/api/routes/bots.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        per_page=per_page,
+        pages=pages,
+    )
+
+
+@router.get(
+    "/{bot_id}",
+    response_model=BotResponse,
+    responses={404: {"model": BotNotFoundError}},
+)
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: **Bug: 404 response has double-nested `detail` key.** The `HTTPException` already wraps its `detail` argument under a `"detail"` key in the JSON response. By passing `detail={"detail": "Bot not found", "id": bot_id}`, the actual response body becomes `{"detail": {"detail": "Bot not found", "id": "..."}}`. The test on line 399 of test_bots.py confirms this odd structure (`data["detail"]["detail"]`).

This is a confusing API contract for consumers. Change to:
```python
raise HTTPException(
    status_code=404,
    detail=f"Bot {bot_id} not found",
)
```
Or if you want structured error info, use a flat dict without re-nesting `detail`:
```python
raise HTTPException(
    status_code=404,
    detail={"message": "Bot not found", "id": bot_id},
)
```
And update `BotNotFoundError` schema + tests accordingly.

File: src/jm_api/db/session.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+from functools import lru_cache
+
+from sqlalchemy import create_engine
+from sqlalchemy.engine import Engine
+from sqlalchemy.orm import Session, sessionmaker
+
+from jm_api.core.config import get_settings
+
+
+@lru_cache
```
Comment (miller46bot) [inline src/jm_api/db/session.py]: **`@lru_cache` on `get_engine` and `get_session_factory` creates global singletons that are hard to override and leak across tests.** While `reset_db_state()` exists to clear caches, this pattern is fragile - any code that calls `get_engine()` before `reset_db_state()` will pin the cached value for the process lifetime, and `lru_cache` is not thread-safe for initialization.

The test suite works around this by overriding the `get_db` dependency entirely (never calling `get_engine`/`get_session_factory`), which means the `reset_db_state()` function has no test coverage and the production path through these cached functions is untested.

Consider using a simple module-level `_engine: Engine | None = None` pattern, or better yet, store these on the `Settings`/app state so they naturally follow the app lifecycle rather than being process-global.

File: src/jm_api/api/routes/bots.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    last_update_at_after: datetime | None = Query(default=None),
+    last_update_at_before: datetime | None = Query(default=None),
+    last_run_at_after: datetime | None = Query(default=None),
+    last_run_at_before: datetime | None = Query(default=None),
+    db: Session = Depends(get_db),
+) -> BotListResponse:
+    """List all bots with filtering, pagination, and sorting."""
+    # Cap per_page at 100
+    per_page = min(per_page, 100)
+
```
Comment (miller46bot) [inline src/jm_api/api/routes/bots.py]: **Silent cap of `per_page` at 100 is surprising API behavior.** A client requesting `per_page=200` gets back `per_page: 100` in the response with no indication that their value was modified. This violates the principle of least surprise. Consider either:
1. Rejecting values > 100 with a 422 (add `le=100` to the Query validator), or
2. Adding a header or field to indicate the value was capped.

Option 1 is simpler and more explicit. You already have `ge=1` validation - adding `le=100` is consistent.

File: src/jm_api/db/session.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (miller46bot) [reply src/jm_api/db/session.py]: Still unresolved. `get_engine` and `get_session_factory` still use `@lru_cache`. The `reset_db_state()` helper was added (useful for tests), but the core concern remains: `lru_cache` is not thread-safe during initialization, and these are global mutable singletons that persist for process lifetime. For a production FastAPI app serving concurrent requests, consider using FastAPI's lifespan event to create the engine once at startup.

File: src/jm_api/api/routes/bots.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+from jm_api.db.session import get_db
+from jm_api.models.bot import Bot
+from jm_api.schemas.bot import BotListResponse, BotNotFoundError, BotResponse
+
+router = APIRouter(prefix="/bots", tags=["bots"])
+
+
+@router.get("", response_model=BotListResponse)
+def list_bots(
```
[the remark this answers is not in the document]
Comment (miller46bot) [reply src/jm_api/api/routes/bots.py]: Partially resolved. The return type annotation `-> BotListResponse` is technically correct and consistent. However, the suggestion to extract query-building logic into a helper/service function for testability and maintainability was not addressed — the function is still a single ~40-line block handling filtering, counting, pagination, and serialization. This is acceptable for now but will become harder to maintain as filters grow. Marking as resolved since the annotation concern was the primary issue.

File: src/jm_api/api/routes/bots.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        query = query.where(Bot.last_run_at >= last_run_at_after)
+    if last_run_at_before is not None:
+        query = query.where(Bot.last_run_at <= last_run_at_before)
+
+    # Get total count
+    count_query = select(func.count()).select_from(query.subquery())
+    total = db.execute(count_query).scalar() or 0
+
+    # Apply sorting and pagination (secondary sort by id for deterministic order)
+    query = query.order_by(Bot.create_at.desc(), Bot.id.desc())
```
[the remark this answers is not in the document]
Comment (miller46bot) [reply src/jm_api/api/routes/bots.py]: Still unresolved. The count query still uses `select(func.count()).select_from(query.subquery())` at line 63, which wraps the filtered query as a subquery for counting. For large tables this is slower than applying the same WHERE clauses to a direct `select(func.count()).select_from(Bot).where(...)`. Consider extracting filter-building into a shared function used by both the data and count queries.

File: src/jm_api/core/config.py [hunk 6]
```diff
@@ -2,16 +2,33 @@
...
     app_name: str = Field(default="jm-api")
     app_version: str = Field(default="0.1.0")
     environment: str = Field(default="development")
     debug: bool = Field(default=False)
 
+    database_url: str = Field(default="sqlite:///./dev.db")
+
+    @model_validator(mode="after")
+    def validate_database_url_for_environment(self) -> "Settings":
+        """Validate database_url is appropriate for the environment."""
```
[the remark this answers is not in the document]
Comment (miller46bot) [reply src/jm_api/core/config.py]: Still unresolved. `database_url` still defaults to `sqlite:///./dev.db` — a relative path that creates a file wherever CWD happens to be. The production validator was a good addition, but the dev default is still fragile. Consider using an absolute path or requiring explicit configuration.

**Review by miller46bot — CHANGES_REQUESTED** [review]
Most prior issues have been addressed — rollback handling was added to `get_db`, the conftest was refactored to use a proper fixture factory, tests no longer import from conftest directly, and `bot_id` now has a `Path` regex constraint. However, the `@lru_cache` singleton concern, the subquery-based count performance issue, and the relative `sqlite:///./dev.db` default path remain unresolved.

**Review by miller46bot — CHANGES_REQUESTED** [review]
The PR implements a solid read-only Bot API with good test coverage, proper SQL wildcard escaping, and sensible pagination. However, there are several issues that should be addressed before merging.

**Review by miller46bot — APPROVED** [review]
No prior issues were listed, so there is nothing to evaluate as unresolved. The PR is approved.

**Optional suggestions (non-blocking):**

1. **`get_db` session lifecycle**: The `get_db` dependency yields a session and rolls back on exception, which is correct. Consider whether you also want to call `session.commit()` on the happy path, or if you prefer the current read-only pattern where the route never commits (appropriate for read-only endpoints).

2. **`conftest.py` session-scoped `set_test_env`**: The session-scoped `monkeypatch_session` fixture using `_pytest.monkeypatch.MonkeyPatch` directly is a known pattern but is reaching into pytest internals. The `pytest-env` plugin or a simple `os.environ` set/restore in a session fixture would be more conventional.

3. **`BotNotFoundError` schema**: The `responses={404: {"model": BotNotFoundError}}` in the OpenAPI spec defines `message` and `id` at top level, but the actual 404 response wraps them under `detail` (FastAPI's `HTTPException` behavior). The OpenAPI docs may not accurately reflect the real response shape.

4. **`pages` calculation when empty**: When `total=0`, `pages` is set to `0`. Some APIs return `pages=1` for consistency (page 1 exists, it's just empty). Minor design choice.
